staging/src/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/queueset/queueset_test.go

```diff
@@ -993,6 +993,79 @@ func TestSelectQueueLocked(t *testing.T) {
 	}
 }
 
+func TestFinishRequestLocked(t *testing.T) {
+	tests := []struct {
+		name         string
+		workEstimate fcrequest.WorkEstimate
+	}{
+		{
+			name: "request has additional latency",
+			workEstimate: fcrequest.WorkEstimate{
+				Seats:             10,
+				AdditionalLatency: time.Minute,
+			},
+		},
+		{
+			name: "request has no additional latency",
+			workEstimate: fcrequest.WorkEstimate{
+				Seats: 10,
+			},
+		},
+	}
+
+	metrics.Register()
+	for _, test := range tests {
+		t.Run(test.name, func(t *testing.T) {
+			metrics.Reset()
+
+			now := time.Now()
+			clk, _ := testeventclock.NewFake(now, 0, nil)
+			qs := &queueSet{
+				clock:   clk,
+				obsPair: newObserverPair(clk),
+			}
+			queue := &queue{
+				requests: newRequestFIFO(),
+			}
+			r := &request{
+				qs:           qs,
+				queue:        queue,
+				workEstimate: test.workEstimate,
+			}
+
+			qs.totRequestsExecuting = 111
+			qs.totSeatsInUse = 222
+			queue.requestsExecuting = 11
+			queue.seatsInUse = 22
+
+			var (
+				queuesetTotalRequestsExecutingExpected = qs.totRequestsExecuting - 1
+				queuesetTotalSeatsInUseExpected        = qs.totSeatsInUse - int(test.workEstimate.Seats)
+				queueRequestsExecutingExpected         = queue.requestsExecuting - 1
+				queueSeatsInUseExpected                = queue.seatsInUse - int(test.workEstimate.Seats)
+			)
+
+			qs.finishRequestLocked(r)
+
+			// as soon as AdditionalLatency elapses we expect the seats to be released
+			clk.SetTime(now.Add(test.workEstimate.AdditionalLatency))
```

Comment:
Why not run clk.Run(nil)

[but then Mike`s comment about proper initialization of QS is a valid one]

Reply:
> `newTestableQueueSetFactory` is intended to be the start of how a test func creates a queueSet.
> Note that when doing goroutine counting, you need to use a promiseFactoryFactory that participates.


@MikeSpreitzer  do you mean this initialization pattern?
```
	clk, counter := testeventclock.NewFake(now, 0, nil)
	qsf := newTestableQueueSetFactory(clk, countingPromiseFactoryFactory(counter))
	qCfg := fq.QueuingConfig{
		Name:             "TestTotalRequestsExecutingWithPanic",
		DesiredNumQueues: 0,
		RequestWaitLimit: 15 * time.Second,
	}
	qsc, err := qsf.BeginConstruction(qCfg, newObserverPair(clk))
	if err != nil {
		t.Fatal(err)
	}
	qs := qsc.Complete(fq.DispatchingConfig{ConcurrencyLimit: 1})
	counter.Add(1) // account for the goroutine running this test
```

Reply:
> Why not run clk.Run(nil)

@wojtek-t for the test at hand, `clk.SetTime` is more concise in expressing the expectation that the event function should be executed immediately after `AdditionalLatency` elapses. So this is a verification step.

Alternatively, we can achieve this with `clk.Run(nil)` too:
```
clk.Run(nil)
get the time when the event fired and assert that it happened after AdditionalLatency elapsed
```

I think `SetTime` does the job concisely, thoughts?

Reply:
In my opinion, using `SetTime` to the precise time needed is sufficient for a test of `finishRequestLocked`.  Using `clk.Run(nil)` would also be OK, and adding a test that `Run` advanced the fake time to the exact correct value would be better.